libkbfs/mdserver_tlf_storage.go

```diff
@@ -209,88 +210,108 @@ func (s *mdServerTlfStorage) checkGetParamsLocked(
 	return nil
 }
 
-func (s *mdServerTlfStorage) getRangeLocked(ctx context.Context,
-	kbpki KBPKI, bid BranchID, start, stop MetadataRevision) (
+func (s *mdServerTlfStorage) getRangeLocked(
+	currentUID keybase1.UID, deviceKID keybase1.KID,
+	bid BranchID, start, stop MetadataRevision) (
 	[]*RootMetadataSigned, error) {
-	err := s.checkGetParamsLocked(ctx, kbpki, bid)
+	err := s.checkGetParamsLocked(currentUID, deviceKID, bid)
 	if err != nil {
 		return nil, err
 	}
 
-	var rmdses []*RootMetadataSigned
-	startKey, err := s.getMDKey(start, bid)
+	j, err := s.getBranchJournalLocked(bid)
 	if err != nil {
-		return rmdses, MDServerError{err}
+		return nil, err
 	}
-	stopKey, err := s.getMDKey(stop+1, bid)
+
+	realStart, mdIDs, err := j.getRange(start, stop)
 	if err != nil {
-		return rmdses, MDServerError{err}
+		return nil, err
 	}
-
-	iter := s.idDb.NewIterator(&util.Range{Start: startKey, Limit: stopKey}, nil)
-	defer iter.Release()
-	for iter.Next() {
-		id, err := MdIDFromBytes(iter.Value())
+	var rmdses []*RootMetadataSigned
+	for i, mdID := range mdIDs {
+		expectedRevision := realStart + MetadataRevision(i)
+		rmds, err := s.getMDLocked(mdID)
 		if err != nil {
 			return nil, MDServerError{err}
 		}
-
-		rmds, err := s.getMDLocked(id)
-		if err != nil {
-			return rmdses, MDServerError{err}
+		if expectedRevision != rmds.MD.Revision {
+			panic(fmt.Errorf("expected revision %v, got %v",
+				expectedRevision, rmds.MD.Revision))
 		}
 		rmdses = append(rmdses, rmds)
 	}
-	if err := iter.Error(); err != nil {
-		return rmdses, MDServerError{err}
-	}
 
 	return rmdses, nil
 }
 
+func (s *mdServerTlfStorage) isShutdownLocked() bool {
+	return s.branchJournals == nil
+}
+
 // All functions below are public functions.
 
 var errMDServerTlfStorageShutdown = errors.New("mdServerTlfStorage is shutdown")
 
-func (s *mdServerTlfStorage) getForTLF(ctx context.Context,
-	kbpki KBPKI, bid BranchID) (*RootMetadataSigned, error) {
+func (s *mdServerTlfStorage) journalLength(bid BranchID) (uint64, error) {
 	s.lock.RLock()
 	defer s.lock.RUnlock()
 
-	if s.isShutdown {
+	if s.isShutdownLocked() {
+		return 0, errMDServerTlfStorageShutdown
+	}
+
+	j, err := s.getBranchJournalLocked(bid)
```

Comment:
This can modify `s.branchJournals`, but you've only taken a read lock.  Do we want to bail out early without creating the journal if it doesn't already exist?